### ISSUE TITLE
model convertion from optimized pb to quantized pb model fail after quantization-aware training

### ISSUE THREAD

**Issue by yangshunDragon** [ildoonet/tf-pose-estimation#473]
Hello,

I use **mobilenet_v2_1.4** as backbone to retrain the model with **quantization-aware training**, and I get the ckpt model successfully. The I start to convert the model from step to step as below:

1. run run_checkpoint.py to generate graph.pb, succeed !
python3 run_checkpoint.py --model=mobilenet_v2_1.4  --resize=432x368  --quantize

2. get frozen pb model,  succeed !
python3 freeze_graph.py      \
--input_graph=/path/to/graph.pb      \
--input_checkpoint=/path/to/ckpt-model      \
--output_node_names=Openpose/concat_stage7      \
--output_graph=/path/to/frozen_pose_m2.pb

3. Perform optimization on the frozen model via graph transform,  succeed !
bazel-bin/tensorflow/tools/graph_transforms/transform_graph \
    --in_graph=/path/to/frozen_pose_m2.pb \
    --out_graph=/path/to/optimized_pose_m2.pb \
    --inputs='image:0' \
    --outputs='Openpose/concat_stage7:0' \
    --transforms='
    strip_unused_nodes(type=quint8, shape="1,432,368,3")
    fold_old_batch_norms
    fold_batch_norms
    fold_constants(ignore_errors=False)
    remove_nodes(op=Identity, op=CheckNumerics)'

4. Quantize neural network in 8 bit 
bazel-bin/tensorflow/tools/graph_transforms/transform_graph \
  --in_graph=/path/to/optimized_pose_m2.pb \
  --out_graph=/path/to/quantM2/quantized_pose_m2.pb \
  --inputs='image' \
  --outputs='Openpose/concat_stage7:0' \
  --transforms='add_default_attributes
    strip_unused_nodes(type=quint8, shape="1,432,368,3")
    fold_old_batch_norms
    fold_batch_norms
    fold_constants(ignore_errors=true)
    remove_nodes(op=Identity, op=CheckNumerics)
    quantize_weights
    quantize_nodes
    sort_by_execution_order'

While the step **fail** with below errors reported:
(base) [tclsz@localhost tensorflow-r1.13]$ sh 4optimized_pb_2_quant.sh 
2019-06-06 14:11:58.704848: I tensorflow/tools/graph_transforms/transform_graph.cc:317] Applying add_default_attributes
2019-06-06 14:11:58.734728: I tensorflow/tools/graph_transforms/transform_graph.cc:317] Applying strip_unused_nodes
2019-06-06 14:11:58.759057: I tensorflow/tools/graph_transforms/transform_graph.cc:317] Applying fold_old_batch_norms
2019-06-06 14:11:58.782938: E tensorflow/tools/graph_transforms/transform_graph.cc:263] Beta input to batch norm has bad shape: [48]
2019-06-06 14:11:58.785208: E tensorflow/tools/graph_transforms/transform_graph.cc:264] usage: bazel-bin/tensorflow/tools/graph_transforms/transform_graph
Flags:
        --in_graph=""                           string  input graph file name
        --out_graph=""                          string  output graph file name
        --inputs=""                             string  inputs
        --outputs=""                            string  outputs
        --transforms=""                         string  list of transforms
        --output_as_text=false                  bool    whether to write the graph in text protobuf format

Transforms are:
add_default_attributes
backport_concatv2......

5. Convert pb(frozen/optimized/quantized pb) model to tflite model, **fail**  
bazel run --config=opt tensorflow/lite/toco:toco -- \
--input_file=/path/to/frozen_pose_m2.pb \
--output_file=/path/to/openpose_m2_uint8.tflite \
--input_shapes=1,432,368,3 \
--input_arrays='image' \
--output_arrays='Openpose/concat_stage7'  \
--inference_type=QUANTIZED_UINT8  \
--mean_values=128 \
--std_values=128 \
--change_concat_input_ranges=false \
--allow_custom_ops

Error reported as below:
INFO: Analysed target //tensorflow/lite/toco:toco (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //tensorflow/lite/toco:toco up-to-date:
  bazel-bin/tensorflow/lite/toco/toco
INFO: Elapsed time: 0.303s, Critical Path: 0.00s, Remote (0.00% of the time): [queue: 0.00%, setup: 0.00%, process: 0.00%]
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
INFO: Running command line: bazel-bin/tensorflow/lite/toco/toco '--input_file=/home-ex/tclsz/luojp/tensorflow-r1.13/quantM2/frozen_pose_m2.pb' '--output_file=/home-ex/tclsz/luojp/tensorflow-r1.13/quantM2/openpose_m2_uint8.tflite' '--input_shapes=1,432,368,3' '--input_arrays=image' '--output_arrays=Openpose/concat_stage7' '--inference_type=QUANTIZINFO: Build completed successfully, 1 total action
2019-06-06 14:13:19.603625: I tensorflow/lite/toco/import_tensorflow.cc:1324] Converting unsupported operation: Size
2019-06-06 14:13:19.611983: I tensorflow/lite/toco/import_tensorflow.cc:193] Unsupported data type in placeholder op: 2
2019-06-06 14:13:19.612015: I tensorflow/lite/toco/import_tensorflow.cc:193] Unsupported data type in placeholder op: 2
2019-06-06 14:13:19.612027: I tensorflow/lite/toco/import_tensorflow.cc:193] Unsupported data type in placeholder op: 2
2019-06-06 14:13:19.612087: I tensorflow/lite/toco/import_tensorflow.cc:1324] Converting unsupported operation: Size
......
......
2019-06-06 14:18:06.064321: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] Before Removing unused ops: 3122 operators, 4647 arrays (0 quantized)
2019-06-06 14:18:06.347391: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] Before general graph transformations: 3122 operators, 4647 arrays (0 quantized)
2019-06-06 14:18:06.850112: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] After general graph transformations pass 1: 1060 operators, 1707 arrays (1 quantized)
2019-06-06 14:18:06.935858: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] Before pre-quantization graph transformations: 1060 operators, 1707 arrays (1 quantized)
2019-06-06 14:18:06.965974: I tensorflow/lite/toco/graph_transformations/graph_transformations.cc:39] After pre-quantization graph transformations pass 1: 854 operators, 1501 arrays (1 quantized)
2019-06-06 14:18:06.998686: F tensorflow/lite/toco/tooling_util.cc:1702] Array MobilenetV2/Conv/Conv2D, which is an input to the (Unsupported TensorFlow op: Size) operator producing the output array MobilenetV2/Conv/BatchNorm/Size, is lacking min/max data, which is necessary for quantization. If accuracy matters, either target a non-quantized output format, or run quantized training with your model from a floating point checkpoint to change the input graph to contain min/max information. If you don't care about accuracy, you can pass --default_ranges_min= and --default_ranges_max= for easy experimentation.
5pb_2_tflite.sh: line 24:  3031 Aborted                 (core dumped) bazel run --config=opt tensorflow/lite/toco:toco -- --input_file=$OUTPUT_DIR/frozen_pose_m2.pb --output_file=$OUTPUT_DIR/openpose_m2_uint8.tflite --input_shapes=1,432,368,3 --input_arrays='image' --output_arrays='Openpose/concat_stage7' --inference_type=QUANTIZED_UINT8 --mean_values=128 --std_values=128 --change_concat_input_ranges=false --allow_custom_ops

My step 4 and 5 fail, I do not know what is wrongwith my covertion. Maybe I make the wrong config of (--transforms="" ) when using transform_graph. Anybody know how to solve this problem? Many thanks in advance!  
@ildoonet @toshiy104  @zjbaby

**Comment by yangshunDragon** [comment]
BTW, I use quantization-aware training and those quantization nodes are successfully added. I test the converted frozen pb and optimized pb model, they could generate right outputs. I am just very confused about how to get quantized pb model as @ildoonet described in https://github.com/ildoonet/tf-pose-estimation/blob/master/etcs/training.md. And I am wondering the correct way to covert quantized pb model to tflite model.

**Comment by caiya55** [comment]
Hi, I am also working on the quantization-aware training, but I am still on the first step of training the quantization-aware based model, my goal is training the model and implement it to the Coral device. 
I follow the [quantization](https://www.tensorflow.org/lite/convert/quantization) code to convert the .pb file to tflite file. 
the code looks like this
![image](https://user-images.githubusercontent.com/33747685/59015232-17978a00-8871-11e9-8ec4-1fc41c1bab3b.png)

And then I upload the tflite to the [coral device website] (https://coral.withgoogle.com/web-compiler/)
Then I get the tflite file and run in the coral. 

I use the released mobilenet_v2_small.pb file (this is not quantized model) and it works with low accuray. So I'm thinking of training the model again and do the coverting again.

And my experience is that the mobilenet_v2_small model works but the mobilenet_thin model is not working.
Oh, if this is your personal project, I will appreciate that if you could share your freezed graph so I can just skip the training process. My machine is not good so I have to fix the environment bugs when I try to train the model. It takes several weeks already.

**Comment by yangshunDragon** [comment]
Hi @caiya55  I am also need to use Coral board to verify how fast TPU could be. For now, I do not have to train a model with high accuracy. I just initialize a ckpt model(no fine-tune) and then freeze. If you still need it, you could add my wechat: yangshun_dragon. Since TPU support only quantization-aware trained model, Have you converted quantization-aware trained model to quantized pb model and tflite model successfully?

**Comment by caiya55** [comment]
Ok, I add your wechat already. No actually I have not converted the quantized pb model before. Did you use the train.py file to generate the ckpt file in tf_pose folder ?
